In [47]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [48]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [49]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [50]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [51]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_market_data)

In [52]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(scaled_data, columns=df_market_data.columns, index=df_market_data.index)

# Copy the crypto names from the original DataFrame
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coin_id column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display the scaled DataFrame
df_market_data_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [53]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 12))

In [54]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model = KMeans(n_clusters=i, random_state=0)
    # 2. Fit the model to the data using `df_market_data_scaled`
    model.fit(df_market_data_scaled)
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(model.inertia_)

In [55]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k,
    "inertia": inertia
}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve = pd.DataFrame(elbow_data)


In [56]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k
df_elbow_curve.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k,
    width=800,
    height=400
)



:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [57]:
# Initialize the K-Means model using the best value for k (4)
model = KMeans(n_clusters=4, random_state=0)


In [58]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_market_data_scaled)


KMeans(n_clusters=4, random_state=0)

In [59]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
df_market_data_scaled["class"] = model.predict(df_market_data_scaled)


# Print the resulting array of cluster values
print(df_market_data_scaled["class"].unique())

[2 0 1 3]


In [60]:
# Create a copy of the scaled DataFrame
df_market_data_scaled_clustered = df_market_data_scaled.copy()

In [61]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_market_data_scaled_clustered["class"] = model.labels_


# Display the copy of the scaled DataFrame
df_market_data_scaled_clustered.head()



price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  class  
coin_id                                          
bitcoin                        -0.251637      2  
ethereum                       -0.199352      2  
tether                         -0.282061      0  
ripple                         -0.295546      0  
bitcoin-cash                   -0.270317      2

In [62]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point
df_market_data_scaled_clustered.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="class",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis",
    width=800,
    height=400
)


:NdOverlay   [class]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [63]:
# Create a PCA model instance and set `n_components=3`
pca = PCA(n_components=3)


In [64]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components
market_data_pca = pca.fit_transform(df_market_data_scaled)

#View the scaled PCA data
market_data_pca


array([[ 1.19408229e+00, -9.02074016e-01, -5.85337627e-01],
       [ 1.00970605e+00, -6.63584379e-01, -1.13695032e+00],
       [-7.22535913e-01, -3.07632409e-01,  7.21813300e-01],
       [-7.48266457e-01, -3.32378949e-01,  5.58478692e-01],
       [ 2.25853853e+00, -1.82696614e+00, -1.37816649e+00],
       [ 1.71918788e+00, -9.93613477e-01, -6.63561320e-01],
       [ 1.39640067e+00, -7.21541835e-01, -2.73134093e+00],
       [ 1.35918692e+00, -6.51114388e-01, -1.48216215e+00],
       [ 8.69062197e-01, -8.28723573e-01, -8.56778699e-01],
       [-7.63186367e-01, -6.19274690e-01,  3.15895319e-01],
       [-1.79599089e+00,  2.50827670e-01,  9.84707174e-01],
       [-7.33772813e-01, -3.10012848e-01,  7.37214639e-01],
       [-1.02424943e+00, -4.66951081e-01,  4.89503524e-01],
       [ 3.24851369e+00, -9.17857212e-01, -6.54640079e-01],
       [-5.53164102e-01, -3.03202134e-01,  2.13756819e-01],
       [ 1.59464452e-01, -7.02818911e-01, -1.62407310e+00],
       [-1.55926251e+00,  5.07701028e-01

In [65]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component
explained_variance = pca.explained_variance_ratio_

# Print the explained variance
explained_variance



array([0.37269822, 0.32489961, 0.18917649])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is 88.68% (rounded to two decimal places for the purposes of this written answer only )
Total Explained Variance = PC1 + PC2 + PC3
                         = 0.37269822 + 0.32489961 + 0.18917649
                         = 0.88677432

In [66]:
# Create a new DataFrame with the PCA data
df_market_data_pca = pd.DataFrame(
    market_data_pca,
    columns=["PC1", "PC2", "PC3"]
)
# Copy the crypto names from the original scaled DataFrame
df_market_data_pca["coin_id"] = df_market_data_scaled.index

# Set the coin_id column as index
df_market_data_pca = df_market_data_pca.set_index("coin_id")

# Display the scaled PCA DataFrame
df_market_data_pca.head(5)

PC1       PC2       PC3
coin_id                                   
bitcoin       1.194082 -0.902074 -0.585338
ethereum      1.009706 -0.663584 -1.136950
tether       -0.722536 -0.307632  0.721813
ripple       -0.748266 -0.332379  0.558479
bitcoin-cash  2.258539 -1.826966 -1.378166

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [67]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1, 12))


In [68]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
for i in k_pca:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    model_pca = KMeans(n_clusters=i, random_state=0)
    # 2. Fit the model to the data using `df_market_data_pca`
    model_pca.fit(df_market_data_pca)
    # 3. Append the model.inertia_ to the inertia list
    inertia_pca.append(model_pca.inertia_)


In [69]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k": k_pca,
    "inertia": inertia_pca
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve_pca = pd.DataFrame(elbow_data_pca)



In [70]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k
df_elbow_curve_pca.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve with PCA",
    xticks=k_pca,
    width=800,
    height=400
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The k value is the same in both sets of data, but the inertia is lower in the PCA data

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [71]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters=4, random_state=0)

In [72]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=0)

In [73]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
df_market_data_pca["class"] = model_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values
print(df_market_data_pca["class"].unique())

[1 0 2 3]


In [74]:
# Create a copy of the scaled PCA DataFrame
df_market_data_pca_clustered = df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_market_data_pca_clustered["class"] = model_pca.labels_

# Display the copy of the scaled PCA DataFrame
df_market_data_pca_clustered.head()


PC1       PC2       PC3  class
coin_id                                          
bitcoin       1.194082 -0.902074 -0.585338      1
ethereum      1.009706 -0.663584 -1.136950      1
tether       -0.722536 -0.307632  0.721813      0
ripple       -0.748266 -0.332379  0.558479      0
bitcoin-cash  2.258539 -1.826966 -1.378166      1

In [75]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point
df_market_data_pca_clustered.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="class",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis with PCA",
    width=800,
    height=400
)



:NdOverlay   [class]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [78]:
# Composite plot to contrast the Elbow curves
df_elbow_curve.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k,
    width=400,
    height=400
) * df_elbow_curve_pca.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve with PCA",
    xticks=k_pca,
    width=400,
    height=400
)





:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [79]:
# Composite plot to contrast the clusters
df_market_data_scaled_clustered.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="class",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis",
    width=400,
    height=400
) + df_market_data_pca_clustered.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="class",
    hover_cols=["coin_id"],
    title="Cryptocurrency Analysis with PCA",
    width=400,
    height=400
)



:Layout
   .NdOverlay.I  :NdOverlay   [class]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [class]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using PCA for feature reduction in K-Means clustering improves cluster clarity, reduces noise, enhances visualization, and boosts computational efficiency. However, it's crucial to ensure the principal components retain enough variance to accurately represent the original data.